### PR TITLE
add ppo

### PR DESCRIPTION
**Pull request by hughian**
train了一下依然没有收敛

### CODE REVIEW

File: Pommer/sacd/sac.py [hunk 1]
```diff
@@ -33,8 +33,9 @@ def __init__(self, num_inputs, action_space, args):
         if self.policy_type != "Gaussian":
             # Target Entropy = −dim(A) (e.g. , -6 for HalfCheetah-v2) as given in the paper
             if self.automatic_entropy_tuning:
-                self.target_entropy = -np.log((1.0 / action_space.n)) * 0.98
+                self.target_entropy = 0 - action_space.n
```
Comment (zhangli140) [inline Pommer/sacd/sac.py:36]: 这个地方是改entropy的hotfix吗？如果是就标记一下，免得到时候不记得了。

File: Pommer/sacd/sac.py [hunk 2]
```diff
@@ -33,8 +33,9 @@ def __init__(self, num_inputs, action_space, args):
...
                 self.log_alpha = torch.zeros(1, requires_grad=True, device=self.device)
+                self.alpha = self.log_alpha.detach().exp()
```
Comment (zhangli140) [inline Pommer/sacd/sac.py:38]: 这里如果是hotfix，也标记下吧

File: Pommer/sacd/sac.py [hunk 3]
```diff
@@ -117,53 +117,61 @@ def _translate_obs(self, o):
...
 
             qf1_next_target, qf2_next_target = self.critic_target(next_state_batch)
-            min_qf_next_target = pi_ * (torch.min(qf1_next_target, qf2_next_target) - self.alpha * log_pi_)
-            # TODO ? \mean_{a'} Q(s', a') Expected Q?
-            min_qf_next_target = min_qf_next_target.mean(dim=1).unsqueeze(-1)
-            next_q_value = reward_batch + (1 - mask_batch) * self.gamma * min_qf_next_target
+            log_pi_ = log_pi_.gather(dim=1, index=next_actions.view(-1, 1).long())
+            qf1_next_target = qf1_next_target.gather(dim=1, index=next_actions.view(-1, 1).long())  # (batch_size, 1)
+            qf2_next_target = qf2_next_target.gather(dim=1, index=next_actions.view(-1, 1).long())  # (batch_size, 1)
+            min_qf_next_target = (torch.min(qf1_next_target, qf2_next_target) - self.alpha * log_pi_)
```
Comment (zhangli140) [inline Pommer/sacd/sac.py:138]: OK

File: Pommer/run_toy_env.py [hunk 4]
```diff
@@ -0,0 +1,147 @@
+import gym
+import datetime
+# from dqn import DQN
```
Comment (zhangli140) [inline Pommer/run_toy_env.py:3]: 这是试了一下DQN吗？效果怎样，上面那个DQN是不是就只在这里用了。

File: Pommer/run.py [hunk 5]
```diff
@@ -88,57 +88,49 @@
...
             if not env._agents[0].is_alive:
                 done = True
+
             episode_step += 1
             total_step += 1
-            episode_reward += rewards[0]
-            obs_ = sac_agent._translate_obs(state_[0])
+            episode_reward += reward
+            if done and len(ppo_agent.buffer) > args.batch_size:
+                ppo_agent.learn(memory, args.batch_size, writer=writer)
```
Comment (zhangli140) [inline Pommer/run.py:120]: 这个有没有输出的各种图啊，我想看看，然后分析一下再给你feedback

File: Pommer/eda.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    o[12] = obs['bomb_blast_strength']
+    o[13] = obs['bomb_life']
+    o[14] = obs['bomb_moving_direction']
+    o[15] = obs['flame_life']
+
+    vec = np.array([obs['ammo'], obs['can_kick'], obs['blast_strength'], obs['step_count']/800], dtype=np.float32)
+    return o, vec
+
+
+def compute_reward(obs, obs_, rewards):
```
Comment (zhangli140) [inline Pommer/eda.py:20]: Oh, 发现了，这是有shaping的！

File: Pommer/ppo/ppo.py [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+                 ppo_update_time=10,
+                 buffer_capacity=1000,
+                 batch_size=32):
+        super(PPO, self).__init__()
+        self.gamma = gamma
+        self.log_interval = log_interval
+        self.clip_param = clip_param
+        self.max_grad_norm = max_grad_norm
+        self.ppo_update_time = ppo_update_time
+        self.buffer_capacity = buffer_capacity  # TODO we want a buffer outside this class
```
Comment (zhangli140) [inline Pommer/ppo/ppo.py:37]: 有空可以专门写个buffer类，复用一下

File: Pommer/ppo/ppo.py [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+
+                ratio = torch.exp(torch.log(action_prob) - torch.log(old_action_prob[index]))
+                surr1 = ratio * advantage
+                surr2 = torch.clamp(ratio, 1 - self.clip_param, 1 + self.clip_param) * advantage
+
+                # update actor network
+                action_loss = -torch.min(surr1, surr2).mean()
+
+                writer.add_scalar('loss/action_loss', action_loss, global_step=self.training_step)
+                self.actor_optimizer.zero_grad()
```
Comment (zhangli140) [inline Pommer/ppo/ppo.py:135]: 这个过程写得不错！和公式完全一致

File: Pommer/ppo/ppo.py [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+                self.critic_optimizer.zero_grad()
+                value_loss.backward()
+                nn.utils.clip_grad_norm_(self.critic_net.parameters(), self.max_grad_norm)
+                self.critic_optimizer.step()
+                self.training_step += 1
+
+        del self.buffer[:]  # clear experience
+
+
+def run_cart_pole(render=False):
```
Comment (zhangli140) [inline Pommer/ppo/ppo.py:152]: 这东西有空拿出去吧，不过暂时放在这也行，不是啥急事

**Review by zhangli140 — COMMENTED** [review]
有空把loss,reward发我下，我要琢磨琢磨，我说的是走势图

File: Pommer/sacd/sac.py (hunk 1 above)
Comment (hughian) [reply Pommer/sacd/sac.py:36]: 参照了一下不同的做法，目标entropy的值设定成了`-n`

File: Pommer/sacd/sac.py (hunk 2 above)
Comment (hughian) [reply Pommer/sacd/sac.py:38]: 好的

File: Pommer/run_toy_env.py (hunk 4 above)
Comment (hughian) [reply Pommer/run_toy_env.py:3]: 对，dqn只用来试一下run_toy_env的循环有没有写错，没有放到Pommerman上试。

File: Pommer/run.py (hunk 5 above)
Comment (hughian) [reply Pommer/run.py:120]: 有的

File: Pommer/eda.py (hunk 6 above)
Comment (hughian) [reply Pommer/eda.py:20]: 嗯，这里是我写的

File: Pommer/ppo/ppo.py (hunk 7 above)
Comment (hughian) [reply Pommer/ppo/ppo.py:37]: 嗯，这里是我标记的，DQN和SAC用的有一个（state, a, reward, next_state, done），PPO不需要next_state, 所以没有用这个。

File: Pommer/ppo/ppo.py (hunk 9 above)
Comment (hughian) [reply Pommer/ppo/ppo.py:152]: 好的。

File: Pommer/run.py (hunk 5 above)
Comment (hughian) [reply Pommer/run.py:120]: reward
这个是reward

Comment (hughian) [reply Pommer/run.py:120]: image
这个是value_loss

Comment (hughian) [reply Pommer/run.py:120]: image

**Comment by zhangli140** [comment]
图已收到，我先想想。然后找你讨论。如果是在服务器上，这个周末就放着跑吧。

**Comment by zhangli140** [comment]
看看多跑几个轮数，会达到什么效果。

**Comment by zhangli140** [comment]
如果是自己机器就算了

File: Pommer/ppo/ppo.py (hunk 8 above)
Comment (zhangli140) [reply Pommer/ppo/ppo.py:135]: sorry我看错了，有点不一样的...原版更快...

File: Pommer/run.py (hunk 5 above)
Comment (zhangli140) [reply Pommer/run.py:120]: 还要confirm几个事情：
1. 目前跑了多少个episode? 10000个吗？还是多少，这个需要多跑下。
2. 每次learn的时候，是不是10次sampling mini-batch
3. 然后每个episode是最多800步吧
4. 然后我刚刚对比了你的代码和OpenAI的代码，似乎有两个很重要的不同（也可能是我看错了，如果是的话，这个你需要纠正我）
   a. Paper公式和OpenAI的代码用了资格迹（TD-lambda）这种方法去重新分配了reward/advantage，可以在一定程度上解决reward/advantage稀疏的问题。
   b. Paper公式和OpenAI的代码似乎是按需生成的buffer，这样每次都是新的。而都缓存起来的话，就会导致采样稀疏，这不是reward稀疏，这是另一种稀疏。

**Review by zhangli140 — COMMENTED** [review]
Please check the new comments.

**Comment by zhangli140** [comment]
OpenAI代码：
https://github.com/openai/baselines/blob/master/baselines/ppo1/pposgd_simple.py

File: Pommer/run.py (hunk 5 above)
Comment (hughian) [reply Pommer/run.py:120]: 1. 10k个episode，policy收敛到一个动作上了。
2. 是的，每次learn, sample mini-batch十次，然后学完clear buffer
3. 对的每个episode最多800，到达800环境就会自己停止了（实际train中大概都只不会超过200 step）
4. openai的代码，baselines还是spinningup?
   a. 是不是代码中的Gt ？
   b. 这里每次train完了buffer就清空了。

对于replay buffer，我看到有些文章说，PPO是不需要experience replay的，每次一个episode的trajectory，然后learn.

File: Pommer/ppo/ppo.py [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+
+        for i in range(self.ppo_update_time):
+            for index in BatchSampler(SubsetRandomSampler(range(len(self.buffer))), self.batch_size, False):
+                if self.training_step % 10000 == 0:
+                    self.save_model('PommOneVsOne', suffix=str(self.training_step))
+
+                # with torch.no_grad():
+                Gt_index = Gt[index].view(-1, 1)
+                V = self.critic_net(state[index])
+                delta = Gt_index - V
```
Comment (zhangli140) [inline Pommer/ppo/ppo.py:122]: delta要往前重新分配一下，才是TD-lambda，最简单的，比方这句赢了，以前所有的action全都得给reward为1，不过这个例子太极端。

File: Pommer/run.py (hunk 5 above)
Comment (zhangli140) [reply Pommer/run.py:120]: PPO不是很适合experience replay，太稀疏了。按需从trajectory学才是正常的。

Comment (zhangli140) [reply Pommer/run.py:120]: 10k我觉得不很够，先看看那个资格迹的问题，弄一下，然后周末再跑两天看看（如果用服务器）。

File: Pommer/ppo/ppo.py (hunk 10 above)
Comment (zhangli140) [reply Pommer/ppo/ppo.py:122]: 可以去看一下这个的介绍或者我留下那个链接里的实现吧

File: Pommer/run.py (hunk 5 above)
Comment (zhangli140) [reply Pommer/run.py:120]: a. 不是代码中的Gt，上面批注了一下

Comment (zhangli140) [reply Pommer/run.py:120]: 等等，我仔细看了下，你这个Gt好像是可以的
